validate push and pull request events using the direct API before triggering

doudou opened the pull request:
We regularly have push events to master whose SHA differs from the
one that is actually at HEAD. Filter these events out in order to
avoid triggering in a loop.

doudou committed:
validate push events before triggering from them

We regularly have push events to master whose SHA differs from the
one that is actually at HEAD. Filter these events out in order to
avoid triggering in a loop.

g-arjones commented on lib/autoproj/daemon/github_watcher.rb:
```diff
@@ -105,6 +105,15 @@ def filter_events(events)
+                    push_event.name
+                )
+                branch_state = branches.find { |b| b.branch_name == push_event.branch }
+                return branch_state&.sha == push_event.head_sha
```
Why the explicit return?

doudou replied:
VSCode is being so damn slow on my machine that I more often than not end up using vim ... hence the rubocop violations piling up :(

I don't understand how it became so bad.

g-arjones replied:
Have you tried disabling vscode-solargraph?

doudou replied:
Yes. Actually, I think it might be an issue with vscode-vim.

doudou replied:
Also had to disable the C++ extension.

g-arjones replied:
> Also had to disable the C++ extension.

This one is a huge memory hog if you a lot of packages. Intellisense caches all the symbols it finds and you end up with a very large database

g-arjones commented:
Could you please elaborate on this? I am not sure I understand what you are trying to fix... It seems to me that the goal is to prevent "old" events (pushes whose HEAD doesn't match current HEAD) from triggering but doesn't the daemon guarantee that already by keeping only the most recent event? Moreover, isn't it possible to have push events that add commits in the middle (out-of-date branches), in that case, the push HEAD and the current HEAD would be different but that's definitely an event we are interested in

doudou commented:
This works around the bug we've been hitting over and over again, with the events reporting a certain commit that's actually not HEAD - and the HEAD commit not being reported at all (or being out-of-order, I'm not too sure to be honest). It's harmless for pull requests, but it hurts when it is the mainline branch - since the daemon then restarts building over and over again.

The only downside to it would be in the case where HEAD is never reported by the event trigger - we wouldn't trigger a build in this particular case. I still believe it's better than the current situation, where we end up paying 12h of build time for nothing. Another possibility would be to use the event as a "hint" that something has been pushed, but generate our own "synthetic event" coming from reading the actual head - this way we're sure to trigger in every case where a push has happened.

g-arjones commented:
> This works around the bug we've been hitting over and over again

I thought as much...

I'm just not so sure this really works around that because:

1. The daemon does an `autoproj update --force-reset` on mainline push so, unless --force-reset isn't force resetting, local and remote HEADs are the same (and we do check the local HEAD already);
2. Dropping old events (timestamp-wise) should have the same effect

To be honest, I think what you might be experiencing is related to [this](https://github.com/rock-core/autoproj-daemon/issues/30) instead. If the last push is omitted then the heads will be different and we end up with the loop but this is a Github bug. This patch might help with that but introduces the problems you mentioned so I'm not sure it's the way to go. 

Personally, I would rather create a push event "cache" and use that to prevent processing the same event more than once. 

Thoughts?

g-arjones commented:
Actually, if what you are experiencing right now is related to the issue I mentioned in my previous comment, this patch won't have any effect at all...

doudou commented:
This *is* working around GitHub's behavior. Now, since GH definitely seems not to be interested in fixing the behavior ... The GH bug happens once a week and every time is *killing* our GCloud bill.

This is a much simpler solution than the cache, does not have the downsides of maintaining a cache, and actually verifies the real question ("do we actually have an event ?"). If anything, it seems that GitHub's position on the event streams is "best effort".

I'm implementing the "synthetic event" solution, which does not have the downside of potentially not triggering.

g-arjones commented:
> I'm implementing the "synthetic event" solution, which does not have the downside of potentially not triggering.

:+1:

doudou committed:
use push events only as hints of what might have happened

Instead of filtering push events, we now just use the presence of a
push event as a "hint" that something has been pushed, calling the
actual API to verify which is the commit tip, and generating a
PushEvent object that matches that tip.

This works around a suspected bug in the GH API, where some events
wouldn't show up (https://github.com/rock-core/autoproj-daemon/issues/30)

g-arjones commented:
OK... Looks good :+1: 

Just a tip: I think it would make things simpler if you just overrode the head_sha [here](https://github.com/rock-core/autoproj-daemon/blob/d3774c705016d4b9f1b6477434e7d98275711e8f/lib/autoproj/daemon/github_watcher.rb#L176) (would at least save you the extra #sort_by and #uniq calls). Also, I don't understand the memo thingy, you don't seem to be using it for anything

doudou committed:
use incoming pull request events only as a hint of what is going on

Like the push events, pull request events are sometimes lost. This
causes pull requests to be kept, that are actually closed (and
mayhem whenever we need to restart a daemon)

This commit implements the same scheme than for the push events:
incoming pull request events are used as a hint that "something"
happened on the pull request. The actual processing is done on
the state of the pull request as fetched from the repository APIs.

doudou commented:
> I don't understand the memo thingy, you don't seem to be using it for anything

Was forgotten. It's meant to avoid calling `client.branches` more than once for a given repository.

I've been implementing the same scheme for the pull requests, and indeed I noticed it would be easier to move the whole logic in the `handle_*` methods. I'll test the new code, and update this pull request for you to review (and re-request a review)

doudou commented:
Updated

g-arjones reviewed (commented):
Looks good :+1:
